engine/univops.ml

```diff
@@ -35,79 +35,29 @@ let universes_of_constr env c =
-  | Some d0 ->
-    let d = merge_types d d0 in
-    if d == d0 then up else LMap.add b d up
-  | None -> LMap.add b d up
-
-let add_up a d b graph =
-  let node, graph =
-    try LMap.find a graph, graph
-    with Not_found ->
-      let node = { up = LMap.empty; visited = false } in
-      node, LMap.add a node graph
-  in
-  node.up <- merge_up d b node.up;
-  graph
-
-(* for each node transitive close until you find a non removable, discard the rest *)
-let transitive_close removable graph =
-  let rec do_node a node =
-    if not node.visited
-    then
-      let keepup =
-        LMap.fold (fun b d keepup ->
-            if not (LSet.mem b removable)
-            then merge_up d b keepup
-            else
-              begin
-                match LMap.find b graph with
-                | bnode ->
-                  do_node b bnode;
-                  LMap.fold (fun k d' keepup ->
-                    merge_up (merge_types d d') k keepup)
-                    bnode.up keepup
-                | exception Not_found -> keepup
-              end
-          )
-          node.up LMap.empty
-      in
-      node.up <- keepup;
-      node.visited <- true
-  in
-  LMap.iter do_node graph
-
-let restrict_universe_context (univs,csts) keep =
-  let removable = LSet.diff univs keep in
-  let (csts, rem) =
-    Constraint.fold (fun (a,d,b as cst) (csts, rem) ->
-        if LSet.mem a removable || LSet.mem b removable
-        then (csts, add_up a d b rem)
-        else (Constraint.add cst csts, rem))
-      csts (Constraint.empty, LMap.empty)
-  in
-  transitive_close removable rem;
-  let csts =
-    LMap.fold (fun a node csts ->
-        if LSet.mem a removable
-        then csts
-        else
-          LMap.fold (fun b d csts -> Constraint.add (a,d,b) csts)
-            node.up csts)
-      rem csts
+let is_constrained u v g =
+  let test = [(u,Eq,v);(u,Lt,v);(v,Lt,u);(u,Le,v);(v,Le,u)] in
+  try Some (List.find (UGraph.check_constraint g) test)
+  with _ -> None
+
+let restrict_universe_context (univs, csts) keep =
+  let allunivs = Constraint.fold (fun (u,_,v) all -> LSet.add u (LSet.add v all)) csts univs in
+  let g = UGraph.empty_universes in
+  let g = LSet.fold UGraph.add_universe_unconstrained allunivs g in
+  let g' = UGraph.merge_constraints csts g in
+  let removed = LSet.diff univs keep in
+  let allkept = LSet.diff allunivs removed in
+  let csts, _g = LSet.fold (fun u cstsg ->
+      LSet.fold (fun v (csts,g as cstsg) ->
+          (** Fold through all pairs of kept universes. We avoid
+             recomputing for swapped order pairs. *)
+          if Level.compare u v >= 0 then cstsg
+          else match is_constrained u v g' with
+            | None -> cstsg
+            | Some cst ->
+              if UGraph.check_constraint g cst then cstsg
+              else Constraint.add cst csts, UGraph.enforce_constraint cst g)
+        allkept cstsg)
+      allkept (Constraint.empty,g)
   in
   (LSet.inter univs keep, csts)
```

Comment:
The universes we keep are exactly the same, as indicated by this line not changing. The constraints are not the same.